libraries/AP_LQR_Control/AP_LQR_Control.cpp

```diff
@@ -0,0 +1,419 @@
+        } else {
+            // select the requested radius, or the required altitude scale, whichever is safer
+            return MAX(sea_level_radius * eas2tas_sq, radius);
+        }
+    }
+}
+
+bool AP_LQR_Control::reached_loiter_target(void)
+{
+    return _WPcircle;
+}
+
+/**
+   prevent indecision in our turning by using our previous turn
+   decision if we are in a narrow angle band pointing away from the
+   target and the turn angle has changed sign
+ */
+
+// update L1 control for waypoint navigation
+void AP_LQR_Control::update_waypoint(const struct Location &prev_WP, const struct Location &next_WP, float dist_min)
+{
+
+    struct Location _current_loc;
+    
+    uint32_t now = AP_HAL::micros();
+    float dt = (now - _last_update_waypoint_us) * 1.0e-6f;
+    if (dt > 0.1) {
+        dt = 0.1;
+    }
+    _last_update_waypoint_us = now;
+
+    // Get current position and velocity
+    if (_ahrs.get_position(_current_loc) == false) {
+        // if no GPS loc available, maintain last nav/target_bearing
+        _data_is_stale = true;
+        return;
+    }
+
+    Vector2f _groundspeed_vector = _ahrs.groundspeed_vector();
+
+    // update _target_bearing_cd
+    _target_bearing_cd = get_bearing_cd(_current_loc, next_WP);
+
+    //Calculate groundspeed
+    float groundSpeed = _groundspeed_vector.length();
+    if (groundSpeed < 0.1f) {
+        // use a small ground speed vector in the right direction,
+        // allowing us to use the compass heading at zero GPS velocity
+        groundSpeed = 0.1f;
+        _groundspeed_vector = Vector2f(cosf(get_yaw_rad()), sinf(get_yaw_rad())) * groundSpeed;
+    }
+
+    // Calculate the NE position of WP B relative to WP A
+    Vector2f AB = location_diff(prev_WP, next_WP);
+
+    // Check for AB zero length and track directly to the destination
+    // if too small
+    if (AB.length() < 1.0e-6f) {
+        AB = location_diff(_current_loc, next_WP);
+        if (AB.length() < 1.0e-6f) {
+            AB = Vector2f(cosf(get_yaw_rad()), sinf(get_yaw_rad()));
+        }
+    }
+    AB.normalize();
+
+    // Calculate the NE position of the aircraft relative to WP A
+    Vector2f A_air = location_diff(prev_WP, _current_loc);
+
+    // calculate distance to target track, for reporting
+    _crosstrack_error = (A_air % AB)/(AB.length());
+
+    _crosstrack_error=- _crosstrack_error;
+    
+    //set minimum turn radius equal to twice the ground speed
+    float min_turn_rad = 2*groundSpeed;
+    
+    //limit maximum lateral acceleration and set default if zero
+    if((_max_latacc > (100*groundSpeed*groundSpeed/min_turn_rad)) || (_max_latacc ==0))
+    { 
+        _max_latacc = 100*groundSpeed*groundSpeed/min_turn_rad;
+    }
+    
+    //Caluclate adaptive gains
+    float q1= sqrtf(float((_max_xtrack*0.01)/(fabsf((_max_xtrack*0.01)-_crosstrack_error))));
+    
+    //Calculate the approach velocity towards path
+    float si = RadiansToCentiDegrees(get_yaw_rad())*0.01;
```

Comment:
This should not be yaw from AHRS. For the method, this is the angle of ground speed.
I have changed it to `float si = RadiansToCentiDegrees(atan2f(_groundspeed_vector.y, _groundspeed_vector.x))*0.01;`
Here are the results with 15 m/s with rather small values for settings:
![image](https://user-images.githubusercontent.com/7996972/50400650-04947000-0791-11e9-8b8d-eb5ce87ddc6a.png)

Reply:
This is actually very natural as the ![image](https://user-images.githubusercontent.com/7996972/50401777-cc455f80-0799-11e9-9035-80919f01e1be.png) which is actually the integral term.